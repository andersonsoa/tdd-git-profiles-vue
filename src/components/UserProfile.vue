<template>
  <div class="user-profile__container">
    <img :src="user.avatar_url" class="user-profile__avatar" />
    <h1 class="user-profile__name">{{ user.name }}</h1>
    <p class="user-profile__bio">{{ user.bio }}</p>
  </div>
</template>

<script>
export default {
  name: "UserProfile",
  props: {
    user: {
      require: true,
      type: Object
    }
  }
};
</script>

<style scoped>
.user-profile__container {
  width: 100%;
  height: 500px;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
}
.user-profile__avatar {
  width: 300px;
  height: 300px;
  border-radius: 50%;
  border: 10px solid black;
  box-shadow: 1px 1px 10px #0004;
}
.user-profile__bio {
  width: 80%;
}
</style>
