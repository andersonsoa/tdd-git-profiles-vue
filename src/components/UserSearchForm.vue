<template>
  <div class="user-search__container">
    <input
      v-model="username"
      @keydown.enter="pesquisar"
      class="user-search__input"
      name="username"
      type="text"
    />
    <button
      @click="pesquisar"
      class="user-search__button"
      type="submit"
      name="submit"
    >
      Pesquisar
    </button>
  </div>
</template>

<script>
export default {
  name: "UserSearchForm",
  data: () => ({
    username: ""
  }),
  methods: {
    pesquisar() {
      this.$emit("submitted", this.username);
      this.username = "";
    }
  }
};
</script>

<style scoped>
.user-search__container {
  width: 100%;
  height: 100px;
  display: flex;
  align-items: center;
  justify-content: center;
}
.user-search__input {
  padding: 10px 15px;
  font-size: 20px;
  border: none;
  border-radius: 4px;
  box-shadow: 1px 1px 3px #0009;
}
.user-search__input:focus {
  outline: none;
}
.user-search__button {
  margin-left: 10px;
  padding: 10px 15px;
  height: 43px;
}
.user-search__button:focus {
  outline: none;
}
</style>
