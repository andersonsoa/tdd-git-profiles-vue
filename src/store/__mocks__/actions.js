import userFixture from "../../../tests/unit/fixtures/user";

export const SEARCH_USER = jest.fn().mockResolvedValue(userFixture);
