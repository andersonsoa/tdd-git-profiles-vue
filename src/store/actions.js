import api from "@/api";

export const SEARCH_USER = async ({ commit }, username) => {
  try {
    const res = await api.searchUser(username);

    commit("SET_USER", res);
  } catch (error) {
    return error;
  }
};
