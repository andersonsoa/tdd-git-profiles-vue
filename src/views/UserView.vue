<template>
  <div class="main__container">
    <UserSearchForm @submitted="pesquisar" />
    <UserProfile :user="user" />
  </div>
</template>

<script>
import { mapActions, mapState } from "vuex";

export default {
  name: "UserView",

  computed: {
    ...mapState({ user: "user" })
  },
  components: {
    UserProfile: () => import("@/components/UserProfile"),
    UserSearchForm: () => import("@/components/UserSearchForm")
  },
  methods: {
    ...mapActions({ searchUser: "SEARCH_USER" }),
    async pesquisar(username) {
      this.searchUser(username);
    }
  }
};
</script>

<style scoped>
.main__container {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  height: 640px;
  width: 700px;
  background-color: #fafafa;
  border-radius: 8px;
  box-shadow: 2px 2px 20px #0005;
}
</style>
