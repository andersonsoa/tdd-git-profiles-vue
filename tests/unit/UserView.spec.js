jest.mock("@/store/actions");
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import UserView from "@/views/UserView";
import UserSearchForm from "@/components/UserSearchForm";
import UserProfile from "@/components/UserProfile";

import initialState from "@/store/state";
import * as actions from "@/store/actions";
import userFixture from "./fixtures/user";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("UserView.vue", () => {
  let state;

  beforeEach(() => {
    jest.resetAllMocks();
    state = { ...initialState };
  });

  const build = () => {
    const wrapper = shallowMount(UserView, {
      localVue,
      store: new Vuex.Store({
        actions,
        state
      })
    });
    return {
      wrapper,
      userSearchForm: () => wrapper.findComponent(UserSearchForm),
      userProfile: () => wrapper.findComponent(UserProfile)
    };
  };

  it("o componente renderiza corretamente", () => {
    const { wrapper } = build();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renderiza o componente filho UserSearchForm", () => {
    const { userSearchForm } = build();

    expect(userSearchForm().exists()).toBe(true);
  });

  it("renderiza o componente filho UserProfile", () => {
    const { userProfile } = build();

    expect(userProfile().exists()).toBe(true);
  });

  it("passa um usuario para o componente profile via prop", () => {
    state.user = userFixture;
    const { userProfile } = build();

    expect(userProfile().vm.user).toBe(state.user);
  });

  it("pesquisa um usuario quando quando recebe 'submitted'", () => {
    const expectedUser = "andersonsoa";
    const { userSearchForm } = build();

    userSearchForm().vm.$emit("submitted", expectedUser);

    expect(actions.SEARCH_USER).toHaveBeenCalled();
    expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual(expectedUser);
  });
});
