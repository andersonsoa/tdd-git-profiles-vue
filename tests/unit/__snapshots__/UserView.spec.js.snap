// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView.vue o componente renderiza corretamente 1`] = `
<div class="main__container">
  <!---->
  <!---->
</div>
`;
